src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2052,99 +2057,70 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
   }
 }
 
-// Given an expression (AndX shift mask) or (AndX mask shift),
-// determine if the AndX must always produce zero, because the
-// the shift (x<<N) is bitwise disjoint from the mask #M.
+//------------------------------ Sum & Mask ------------------------------
+
+// Returns a lower bound on the number of trailing zeros in expr.
+static jint AndIL_min_trailing_zeros(const PhaseGVN* phase, const Node* expr, BasicType bt) {
+  expr = expr->uncast();
+  const TypeInteger* type = phase->type(expr)->isa_integer(bt);
+  if (type == nullptr) {
+    return 0;
+  }
+
+  if (type->is_con()) {
+    long con = type->get_con_as_long(bt);
+    return con == 0L ? (type2aelembytes(bt) * BitsPerByte) : count_trailing_zeros(con);
+  }
+
+  if (expr->Opcode() == Op_ConvI2L) {
+    expr = expr->in(1)->uncast();
+    bt = T_INT;
+    type = phase->type(expr)->isa_int();
+  }
+
+  if (expr->Opcode() == Op_LShift(bt)) {
+    const TypeInt* rhs_t = phase->type(expr->in(2))->isa_int();
+    if (rhs_t == nullptr || !rhs_t->is_con()) {
+      return 0;
+    }
+    return rhs_t->get_con() % (type2aelembytes(bt) * BitsPerByte);
+  }
+
+  return 0;
+}
+
+// Given an expression (AndX T+expr mask), determine
+// whether expr is neutral wrt addition under mask
+// and hence the result is always equivalent to (AndX T mask),
 // The X in AndX must be I or L, depending on bt.
-// Specifically, the following cases fold to zero,
+// Specifically, this holds for the following cases,
 // when the shift value N is large enough to zero out
 // all the set positions of the and-mask M.
 //   (AndI (LShiftI _ #N) #M) => #0
 //   (AndL (LShiftL _ #N) #M) => #0
 //   (AndL (ConvI2L (LShiftI _ #N)) #M) => #0
+//   as well as for constant operands:
+//   (AndI (ConI [+-] _ << #N) #M) => #0
```

Comment:
`(AndI (ConI (_ << #N)) #M)`

I think writing like this is clearer, it is confusing talking about signs in bitwise operations. Also, please remove `=> #0` in these.